Training24/Test.c

```diff
@@ -1,14 +1,193 @@
+   };
+   int numTests = sizeof (tests) / sizeof (tests[0]);
+   int* input = malloc (numTests * sizeof (int));
+   int* actualResults = malloc (numTests * sizeof (int));
+   int* pass = malloc (numTests * sizeof (int));
+   if (input == NULL || actualResults == NULL || pass == NULL) {
+      printf ("Memory allocation failed\n");
+      return;
+   }
+   for (int i = 0; i < numTests; ++i) {
+      input[i] = tests[i].input;
+      actualResults[i] = ReverseNumber (tests[i].input);
+      pass[i] = (actualResults[i] == tests[i].expected);
+   }
+   // Call the function to format and print the results
+   FormatResult2 (numTests, input, actualResults, pass);
+   free (input);
+   free (actualResults);
+   free (pass);
+}
+
+void main () {
+   // Run test cases
+   TestPalindrome ();
+   TestReverseNumber ();
+   char choice;
+   while (1) {
+      printf (MAGENTA "Enter the input: " RESET);
+      size_t bufsize = 0;
+      char* buffer = NULL;
+      while (1) {
+         char* temp = realloc (buffer, bufsize + 1);
+         if (temp == NULL) {
+            fprintf (stderr, "Memory allocation failed\n");
+            free (buffer);
+            return;
+         }
+         buffer = temp;
+         int ch = getchar ();
+         if (ch == '\n' || ch == EOF) break;
+         buffer[bufsize++] = (char)ch;
+      }
+      if (bufsize > 4000) printf (RED "The input exceeds the maximum length.\n" RESET);
+      else if (bufsize > 0) {
+         buffer[bufsize] = '\0';
+         // Check for empty string or whitespace
+         if (strlen (buffer) == 0 || strspn (buffer, " \t\r\n") == strlen (buffer)) printf ("Empty string is not a palindrome.\n");
+         else {
+            char* endptr;
+            int num = (int)strtol (buffer, &endptr, 10);
+            if (*endptr == '\0' || *endptr == '\n') {
+               int reversedNum = ReverseNumber (num);
+               printf ((num < 0) ? "Negative numbers are not palindromes.\nReversed Number: %d-\n" :
+                  (reversedNum == -1) ? "Overflow\n" : "Reversed Number: %d \n%s\n",reversedNum, reversedNum == num ? "Palindrome" : "Not a Palindrome");
+            }
+            else {
+               char* temp = malloc (bufsize + 1);
+               if (temp == NULL) {
+                  fprintf (stderr, "Memory allocation failed for temp\n");
+                  free (buffer);
+                  return;
+               }
+               FilterChar (temp, buffer);
+               printf ("%s\n", IsPalindrome (temp) ? "Palindrome" : "Not a Palindrome");
+               free (temp);
+            }
+         }
+      }
+      free (buffer);
+      buffer = NULL;
+      while (1) {
+         printf ("\nDo you want to enter another input? (y/n): ");
+         choice = _getch (); // Get single character input without pressing Enter
+         ClearScreen ();
+         choice = tolower (choice);
+         if (choice == 'y') break;
+         printf (SKYBLUE "Exit the Program" RESET);
+         return;

```

Comment:
The return here causes the function to end instantly, instead of behaving the way you intended it to and letting the loop continue until user enters 'n', it exits after first pass. 
Also, you should check for both 'N' and 'n'.